### ISSUE TITLE
How to turn on Spark

### ISSUE THREAD

**Issue by MadDataScience** [DSPN/amazon-cloudformation-dse#26]
Hi guys,

Getting this running was easy and painless, so well done there. I'm able to connect to the Cassandra cluster and can easily move data in and out of it, so that's great.

However, one of the things that most excited me about the DSE is the integration with Spark. I'm new to DSE in general, so I apologize if this is a stupid question, but I cannot figure out how to turn Spark on in this cluster. The documentation says that nodes with Spark will have a star icon next to it, but nothing (I could find) about how to install Spark onto those nodes. It says that Spark is the default for analytics nodes, but I can't find anything on how to convert a node into an analytics node. Any help on this would be greatly appreciated. 

Thank you in advance,
-Alessandro Gagliardi

**Comment by gmflau** [comment]
Hi Alessandro,

Thanks for using our tool to deploy DSE.  As far as I know, our tool to provision DSE does not turn on Spark by default.  You might follow this guide - https://github.com/DSPN/DataStaxDay/blob/master/labs/Lab%201%20-%20Accessing%20the%20Cluster.md starting  from this "vi /etc/default/dse" command and onwards to turn on Spark.  You will need to do the same for each node of your DSE cluster.

Gilbert

**Comment by cpoczatek** [comment]
Hi Alessandro,

Gilbert's answer is correct. But if you can wait a day or 2 I'm about to push some changes so that the deploy uses DSE's [Lifecycle Manager](https://docs.datastax.com/en/latest-opsc/opsc/LCM/opscLCM.html) which gives a web interface to manage config and other things.

Or if you're feeling adventurous, this is mostly working in the dev branch and I can walk you through it.

Collin

**Comment by MadDataScience** [comment]
Oh, this whole repo is awesome. Thank you. 

Some notes: I had to replace "datastax@" with "ubuntu@" and when I ran `dsetool ring` the first time I got
> Warning: dse_leases has only 1 replicas in the datacenter 'dc0'; a single lease replica failure will prevent the SparkMaster/JobTracker from being elected. You should probably increase the replication factor by running the ALTER KEYSPACE command.
So I entered `cqlsh` and ran `ALTER KEYSPACE dse_leases WITH REPLICATION = { 'class' : 'NetworkTopologyStrategy', 'dc0' : 3 };` which seems to have solved the problem.

Now: PySpark seems to work fine but I like to work in a notebook if possible, so I'm wondering if you have any suggestions.

Update: I was able to get Doan DuyHai's custom build of Zeppelin to work on the cluster with Spark but not PySpark.

**Comment by MadDataScience** [comment]
That sounds great, @cpoczatek . It sounds like the dev branch is mostly working now? I might give that a shot later today. Thanks!

**Comment by cpoczatek** [comment]
@MadDataScience I may have spoken too soon, as I'm not seeing an immediate solution to you problem (spark enabled by default) in the LCM stuff I mentioned. Lemme work on this for a bit.

**Comment by MadDataScience** [comment]
@cpoczatek , the solution in https://github.com/DSPN/DataStaxDay/blob/master/labs/Lab%201%20-%20Accessing%20the%20Cluster.md works well, though it is a bit tedious. It seems like it wouldn't be so hard to come up with a script that would automatically run through the list of nodes, ssh in, make those changes, and restart. Maybe I'll assign that to a student. :)

**Comment by cpoczatek** [comment]
@MadDataScience I pushed the changes to the master branch. If you feel like using them look in [amazon-cloudformation-dse/multidc](https://github.com/DSPN/amazon-cloudformation-dse/tree/master/multidc). Ignore the *multidc* part, you can deploy just one datacenter. These scripts and templates leverage OpsCenter's [Lifecycle Manager](https://docs.datastax.com/en/opscenter/6.0/opsc/LCM/opscLCM.html) for the deployment. This get's you a few things:

- Spark on by default. I set the default config here to have Spark/Solr/Graph on by default. (mixed workloads not recommended for production yada yada yada)

- 99% of DSE's config is exposed in the LCM gui.

- Maybe it makes sense for teaching to give each student a datacenter (with appropriate local replication factors). LCM does support multiple clusters (which would make more sense) but that's not going to work right now.

There's definitely bugs, and the doc could probably be clearer, but if you use it let me know and I'll try to fix stuff.

**Comment by MadDataScience** [comment]
Awesome. Thanks, @cpoczatek . I was planning on have each student deploy their own datacenter (probably using t2.medium to save $$$) so that last bit shouldn't be an issue. Couple questions: 

1. Since multidc can be deployed to just one datacenter, does that mean singledc should be deprecated? Just curious.
2. Wondering if you could elaborate on mixed workloads a bit. I think I follow, but I want to make sure. The idea is that you don't want to be running a Spark job and a Solr query on the same cluster at the same time. But running them in series would be okay, right? And would the problem merely be inefficiency (possibly swapping in and out of memory huge amounts of data unnecessarily) or is there a danger of deadlock?

Thanks!

**Comment by cpoczatek** [comment]
(Sorry for the slow reply, something's wrong with my github notifications...)

1- Yes, eventually singledc/multidc will get merged into a single deploy method. multidc needs to get a little more user friendly before that happens.

2- There aren't any real issues (AFAIK) with running jobs in parallel besides the expected decrease in performance if one of those jobs is heavy. This more comes up in the context of wanting to guarantee data ingest at a certain rate or a fast response time for clients connecting to Cassandra. In those cases you'd usually have a Cassandra-only datacenter and a Spark/Solr/Graph datacenter for analytics.
